Use React Agent for Deepseek Provider and Deepseek V3 Model

ariel-frischer opened the pull request:
- Added langchain-deepseek>=0.1.2
- Use `DeepseekChat` model if `--provider "deepseek" --model "deepseek-chat"`
- Do not use DeepseekChat for openrouter provider - tool calls not supported.
- Use ReAct agent for deepseek v3 + deepseek provider
- Allow `base_state_modifier` to estimate + trim tokens for most ReAct models that are not claude.
-  Tested this out with deepseek-chat, it ran fully without any noticeable issues, went through research + 5 tasks works great.


Commit Changes:
refactor(agent_utils.py): rename sonnet_35_state_modifier to base_state_modifier for clarity and consistency
refactor(agent_utils.py): update state_modifier logic to handle DeepSeek v3 models
refactor(agent_utils.py): improve agent creation logic to support new model types
refactor(anthropic_token_limiter.py): rename sonnet_35_state_modifier to base_state_modifier for better understanding
refactor(model_detection.py): add is_deepseek_v3 function to identify DeepSeek v3 models
refactor(llm.py): enhance create_llm_client to support DeepSeek models and refactor common parameters
test: update tests to reflect changes in state_modifier and model detection logic for better coverage and accuracy

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced DeepSeek V3 integration with enhanced model detection for improved agent configuration.
  - Added new model configuration for "deepseek-chat" to enhance parameter settings.

- **Refactor/Style**
  - Streamlined logic for agent state management and parameter handling, boosting readability and maintainability.
  - Improved formatting and readability across various files.

- **Tests**
  - Expanded test coverage for agent initialization and token limit scenarios, ensuring robust performance.
  - Added new test cases for DeepSeek and OpenAI-compatible model initialization.

- **Chores**
  - Updated project dependencies with a new integration package to support the enhanced functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ai-christianson commented on ra_aid/model_detection.py:
```diff
@@ -86,6 +109,10 @@ def should_use_react_agent(model: BaseChatModel) -> bool:
+    if is_deepseek_v3(model_name) and is_deepseek_model(model):
+        logger.debug(f"Using create_react_agent for DeepSeek model: {model_name}")
+        return True
+
```
We should be getting away from special case logic. The default agent backend for the model should be specified in `models_params.py`.

ai-christianson reviewed (changes requested):
Looks good overall. I like that we're adding the native DeepseekChat api support. Main note: we want to get away from having special cases for specific models, and control things via `models_params`.

Also (somewhat out of scope but worth mentioning): I am observing more consistent behavior with CIAYN rather than `create_react_agent` on dsv3 as well as the new gemini pro 2.5. 

We should test each and pick the one that actually works best, rather than assuming `create_react_agent` always is the best choice. `models_params.py` is where the default for the each model is set (using `"default_backend"`.)

ariel-frischer commented:
> Looks good overall. I like that we're adding the native DeepseekChat api support. Main note: we want to get away from having special cases for specific models, and control things via `models_params`.

Gotcha, I added this because I did want the native API functional tool calling support with deepseek-v3 as I did notice it fails some tool calls on the CiaynAgent. 
- Added to `models_params` in last push and removed special if statement handlers.

> Also (somewhat out of scope but worth mentioning): I am observing more consistent behavior with CIAYN rather than `create_react_agent` on dsv3 as well as the new gemini pro 2.5.
> 
> We should test each and pick the one that actually works best, rather than assuming `create_react_agent` always is the best choice. `models_params.py` is where the default for the each model is set (using `"default_backend"`.)

I agree that we don't actually know which backend agent is best per model. It's possible both backends work fine, or one is better then the other. I think I would give ReAct a slight edge due to native tool calling support. The prompt for react agent also does not need to give that hefty prompt explaining <good code> <bad code> in each message so that should save on some amount of context tokens. Until we can go through comprehensive benchmarking, I think we have to just go by run experience for now. Now that our react agent is working with more providers/models it would be nice if users can specify cli args of `--agent-backend` that should allow for easier testing of different agent backend rather than hardcoding it. 

How about we try pushing this, and just monitor its performance - if users are experiencing any issues it would be a quick one line change to change agent backend in models_params?

ai-christianson commented:
> I think I would give ReAct a slight edge due to native tool calling support.
Check this out, code actually works better, empirically:
https://x.com/AymericRoucher/status/1905313656268882262

> The prompt for react agent also does not need to give that hefty prompt explaining <good code> <bad code> in each message so that should save on some amount of context tokens.

The `create_react_agent` is adding its own prompts, no? You can't just add tools and have an agent.

> Now that our react agent is working with more providers/models it would be nice if users can specify cli args of --agent-backend that should allow for easier testing of different agent backend rather than hardcoding it.

It's not hardcoded, that's why it is called `default_backend` in `models_params` --it is the _default_ for that particular model.

The intent was to add that flag at some point, but the default should be the one that actually works best.

At this point the main thing on the PR is getting rid of the special cases.

As for which should be the backend, I have been having much more consistent behavior on CIAYN backend for all three of these:

- DSv3
- gemini-2.5-pro-exp-03-25
- chatgpt-4o-latest (just released today)
- all smaller models

The specific issue I've been having with `create_react_agent` tests is that the agent will suddenly end the conversation. It will also do things like call research complete even though it did not do any research.

ariel-frischer commented:
> > I think I would give ReAct a slight edge due to native tool calling support.
> 
> Check this out, code actually works better, empirically: https://x.com/AymericRoucher/status/1905313656268882262

Interesting, yea it reminds me of this leaderboard as well: https://gorilla.cs.berkeley.edu/leaderboard.html
Where native function calling is compared to prompt based function calling. I'm not sure if prompt tool calling is referring to "code actions" or generating code to evaluation at runtime. Regardless, its been clear to me that some models like flash 2.0 ran horrible for me in Ciayn agent but ran much better with ReAct due to the native tool calling. Same with sonnet 3.7 so this seems to be on a model by model basis.

> > The prompt for react agent also does not need to give that hefty prompt explaining   in each message so that should save on some amount of context tokens.
> 
> The `create_react_agent` is adding its own prompts, no? You can't just add tools and have an agent.

From my debugging of messages list it seemed like the Ciayn prompt is included in every message for Ciayn but not for ReAct.

> > Now that our react agent is working with more providers/models it would be nice if users can specify cli args of --agent-backend that should allow for easier testing of different agent backend rather than hardcoding it.
> 
> It's not hardcoded, that's why it is called `default_backend` in `models_params` --it is the _default_ for that particular model.
> 
> The intent was to add that flag at some point, but the default should be the one that actually works best.

What I meant is changing the agent backend for some model is a hardcoded change although yes we do have the default ciayn backend setup.

> At this point the main thing on the PR is getting rid of the special cases.

Yes, this is done.

> As for which should be the backend, I have been having much more consistent behavior on CIAYN backend for all three of these:
> 
> * DSv3
> * gemini-2.5-pro-exp-03-25
> * chatgpt-4o-latest (just released today)
> * all smaller models
> 
> The specific issue I've been having with `create_react_agent` tests is that the agent will suddenly end the conversation. It will also do things like call research complete even though it did not do any research.

Ok gotcha I have not seen these issues yet with ReAct agent. If DSV3 is working great in Ciayn we can just leave it for now. There still may be some benefit of the new ChatDeepSeek vs the original ChatOpenAI chat model. It should comply better with any deepseek API changes and have better internal message handlers, etc...

I can revert the backend to Ciayn Agent and keep ChatDeepSeek usage for now your call.
